src/vote_program.rs

```diff
@@ -0,0 +1,143 @@
+
+use bincode::{deserialize, serialize};
+use byteorder::{ByteOrder, LittleEndian};
+use solana_program_interface::account::Account;
+use solana_program_interface::pubkey::Pubkey;
+use std;
+use std::collections::VecDeque;
+use transaction::Transaction;
+
+// Upper limit on the size of the Vote State
+pub const MAX_STATE_SIZE: usize = 1024;
+
+// Maximum number of votes to keep around
+const MAX_VOTE_HISTORY: usize = 32;
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    UserdataDeserializeFailure,
+    InvalidArguments,
+    InvalidUserdata,
+    UserdataTooSmall,
+}
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "error")
+    }
+}
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub struct Vote {
+    // TODO: add signature of the state here as well
+    /// A vote for height tick_height
+    pub tick_height: u64,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum VoteInstruction {
+    NewVote(Vote),
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct VoteProgram {
+    pub votes: VecDeque<Vote>,
+    pub node_id: Pubkey,
+}
+
+pub const VOTE_PROGRAM_ID: [u8; 32] = [
+    6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+];
+
+impl VoteProgram {
+    pub fn check_id(program_id: &Pubkey) -> bool {
+        program_id.as_ref() == VOTE_PROGRAM_ID
+    }
+
+    pub fn id() -> Pubkey {
+        Pubkey::new(&VOTE_PROGRAM_ID)
+    }
+
+    pub fn deserialize(input: &[u8]) -> Result<VoteProgram> {
+        let len = LittleEndian::read_u16(&input[0..2]) as usize;
+
+        if len == 0 {
+            Ok(VoteProgram::default())
+        } else if input.len() < len + 1 {
+            Err(Error::InvalidUserdata)
+        } else {
+            deserialize(&input[2..=len + 1]).map_err(|err| {
+                error!("Unable to deserialize vote state: {:?}", err);
+                Error::InvalidUserdata
+            })
+        }
+    }
+
+    pub fn serialize(self: &VoteProgram, output: &mut [u8]) -> Result<()> {
+        let self_serialized = serialize(self).unwrap();
+
+        if output.len() + 2 < self_serialized.len() {
+            warn!(
+                "{} bytes required to serialize but only have {} bytes",
+                self_serialized.len(),
+                output.len() + 2,
+            );
+            return Err(Error::UserdataTooSmall);
+        }
+
+        let serialized_len = self_serialized.len() as u16;
+        LittleEndian::write_u16(&mut output[0..2], serialized_len);
+        output[2..=serialized_len as usize + 1].clone_from_slice(&self_serialized);
+        Ok(())
+    }
+
+    pub fn process_transaction(
+        tx: &Transaction,
+        instruction_index: usize,
+        accounts: &mut [&mut Account],
+    ) -> Result<()> {
+        if !Self::check_id(&accounts[1].program_id) {
```

Comment:
can you document what you expect accounts 0 and 1 to be?  I was expecting only 1 account, 0, which would store the vote for this tx.